plugins/quota/actor.go

```diff
@@ -0,0 +1,95 @@
+package quota
+
+import (
+	"os"
+	"path"
+
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/glusterd2/xlator"
+	"github.com/gluster/glusterd2/pkg/errors"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type quotadActor struct{}
+
+const (
+	quotaDaemonKey = "enable"
+)
+
+// isQuotadAffectedNode returns true if there are local bricks of volume on
+// which quota is enabled
+func isQuotadAffectedNode(volumes []*volume.Volinfo) bool {
+	for _, v := range volumes {
+		if !isQuotaEnabled(v) {
+			continue
+		} else if v.State != volume.VolStarted {
+			continue
+		} else {
+			bricks := v.GetLocalBricks()
+			if len(bricks) > 0 {
+				return true
+			}
+			continue
+		}
+	}
+	return false
+}
+
+func (actor *quotadActor) Do(v *volume.Volinfo, key, value string, logger log.FieldLogger) error {
+	if key != quotaDaemonKey {
+		return nil
+	}
+	quotadDaemon, err := NewQuotad()
+	if err != nil {
+		return err
+	}
+	// Create pidfile dir if not exists
+	if err := os.MkdirAll(path.Dir(quotadDaemon.pidfilepath),
+		os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+
+	// Create logFiledir dir
+	if err := os.MkdirAll(path.Dir(quotadDaemon.logfilepath),
+		os.ModeDir|os.ModePerm); err != nil {
+		return err
+	}
+
+	volumes, err := volume.GetVolumes()
+	if err != nil {
+		logger.WithError(err).Error("failed to get volumes")
+		return err
+	}
+	// Check for making changes to quotad on this node
+	if !isQuotadAffectedNode(volumes) {
+		// This condition is for disable
+		// TODO: Need to ignore errors where process is already running.
+		daemon.Stop(quotadDaemon, true, logger)
+		logger.Info("quotad is already running")
+	} else {
+		//TODO: Handle ENOENT of pidFile
+		err = daemon.Stop(quotadDaemon, true, logger)
+		if err == errors.ErrPidFileNotFound {
+			logger.Info("quotad stop failed as pidfile missing")
+		} else {
+			logger.Warn("quotad stop failed")
```

Comment:
If you have found that quotad was running during validate step, if pidfile is missing by the time we get here, it's an error.

Reply:
If the pidfile is missing then the daemon is not running, as its a stop whatever was supposed to be done was done. so i think we are fine here.

Reply:
> If the pidfile is missing then the daemon is not running

That assumption may not be true in corner cases (gd2 going down etc). But that's okay, for now.

The validate step would've confirmed that it's running before getting here during a "off" operation. Who would stop quotad after validate step and before we reach this point ? That's the case you described - pidfile not existing because quotad is stopped.

Reply:
If gd2 goes down after validate and before daemon stop, the quotad is still up. and its pidfile holds fine. once it comes up we will still have quotad running. To fix such corner cases we will have to come up with force disable later. correct me if i understood what you are saying.

Reply:
> correct me if i understood what you are saying

> That assumption may not be true in corner cases (gd2 going down etc). But that's okay, for now.

Ignore ^ for now.

> The validate step would've confirmed that it's running before getting here during a "off" operation. Who would stop quotad after validate step and before we reach this point ? That's the case you described - pidfile not existing because quotad is stopped.

I want to know how will the pidfile can be missing when `Do()` is executed  ?

Reply:
During the first quotad start for a node. now no pidfile would be there as quota was never run. and the code flows through the stop and then the start. 
In the stop it will fail as it couldn't find the pidifle and and then it would be actually started in start.